### PR TITLE
refactor(e2e,playwright): input labels to test perf and improve coverage for landing page

### PR DESCRIPTION
**Pull request by Ahmadkashif**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by Ahmadkashif** [comment]
Hi @raisedadead @naomi-lgbt @ojeytonwilliams :wave: 

This PR is a refactor of the old implementation, which had no labelled input.

This PR represents what a typical input labeling effort looks like.

On merging this PR we can:

1. Achieve reliable parallelism support, decreasing test run times by upto 400%, improving CI and local testing experience ( will need changes like these on other playwright test files )
2. Achieve a reliable output by simply following playwright docs. This was not happening in the old implementation.
3. Achieve a much deeper and atomic control over which html element to test.
4. Achieve a framework to support testing all languages within our CI processes.

As always, I'll happy to improve this PR even more.

File: e2e/landing.spec.ts [hunk 1]
```diff
@@ -40,35 +36,78 @@ test.afterAll(async () => {
...
-        'graduates have gotten jobs at tech companies including:'
-    )
-  ).toBeTruthy();
+test('Has 5 brand logos', async () => {
+  const logos = page.getByTestId('brand-logo-container');
+  expect(await logos.nth(0).isVisible());
+  expect(await logos.nth(1).isVisible());
+  expect(await logos.nth(2).isVisible());
+  expect(await logos.nth(3).isVisible());
+  expect(await logos.nth(4).isVisible());
```
Comment (ojeytonwilliams) [inline e2e/landing.spec.ts:61]:
```suggestion
  const logos = page.getByTestId('brand-logo-container').locator('svg');
  await expect(logos).toHaveCount(5);
  for (const logo of await logos.all()) {
    await expect(logo).toBeVisible();
  }
```
This is a little more explicit (i.e. it checks there are 5 logos, not at least 5) and slightly more concise.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks for this refactor, @Ahmadkashif, this looks much better!

Aside from a few stylistic points (below), I was wondering how this impacts parallelism. Is it just that the tests are less flaky or is there more to it than that?

**Comment by Ahmadkashif** [comment]
Hi @ojeytonwilliams 👋 

Lets answer your question on how does this help achieve parallelism?

Consider the old implementation for a second, in order for me to locate and test a div element that has the phrase " Learn to Code " in it, I did the following, page.find(text=Learn to Code).
According to my understanding of this implementation mentioned above, identifying text was flaky on different browsers. The reason, according to my understanding and some googling, led me to a resource stating that browsers have different mechanisms of handling and displaying text. This can potentially take a lot more time for Playwright to navigate to the exact element having the text "Learn to Code".

As a result of this, playwright tests were randomly flaky. I had to disable parallel runs for this flakiness to drop and used hit and trial to come up with the right way of identifying every element. Funny enough you can still see different ways of doing the same thing in the old implementation, they are different because they were failing for webkit or firefox otherwise.

NOTE: Why did I disable parallel runs? Well, according to my understanding, when we dont label input HTML and use other APIs, quite a lot of tests were timing out, for no exact reason. Simply disabling parallelism worked because I think these threads, in some point of execution, waiting for the same resource. That resource was maybe locked by 1 thread, making all others to wait and eventually time out, randomly.

So, what is this refactor about? Well, by labelling input HTML properly, we create key value pairs of these HTML elements which are extremely fast to fetch and work exactly the same on all browsers and these tests work exactly like the Playwright documentation says ( cant say that for the old implementation ). This gives me the power to be 100% certain on the test to never be flaky, in fact, it makes tests 100% reliable, showing 0% random flakiness. 

These refactored tests were isolated and tested for support of parallelism. If all other playwright files test with labelled input, we can be 100% certain that parallelism will work and will be reliable.

**Comment by ojeytonwilliams** [comment]
Awesome, thanks for the detailed breakdown.

So, from my investigations (see https://github.com/freeCodeCamp/freeCodeCamp/pull/51499) I suspect that some of the flakiness can be avoided by using ["web-specific" async matchers](https://playwright.dev/docs/test-assertions#auto-retrying-assertions). In contrast assertions like `expect(await page.isVisible('text=successfully completed')).toBeTruthy();` only tell you if the page is visible when `page.isVisible` resolves. If that text is rendered later, you're out of luck.

That is to say, it's not that checking text is intrinsically flaky; it's more that non-retrying assertions can be.

Finally, bit of a side issue, but given that github action runners only have two cores, I think running the tests on more than one worker will likely _reduce_ performance. After all, at least one is used by the site itself, so more workers will just compete for the same resources.

Having said that, reducing flakiness is something we desperately want, so this PR is valuable even if we completely ignore parallelism!

**Comment by Ahmadkashif** [comment]
I agree with your assessment @ojeytonwilliams.

Retrying a few more times for such flaky tests might work but again it leaves room for flakiness. 

Plus, we are ignoring the fact that properly labelling HTML input and only using getByTestID in writing tests sets a good standard for others to follow. This will mean we can anchor on using getByTestId as a best practice and make contributions much more beginner friendly.

Lastly, while we might not pursue parallelism in CI runs, I'll be at peace knowing that this can still support parallelism, which is enabled by default from Playwright. This can be revisited in the future, but for now, the punchline is, we have removed flakiness and have a good understanding of what works and what might not.

**Review by raisedadead — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
> Retrying a few more times for such flaky tests might work but again it leaves room for flakiness.

Agreed. It's not a silver bullet, but auto-retrying assertions help with a common issue: the DOM mutates after the page has loaded.

> getByTestId as a best practice and make contributions much more beginner friendly.

Yeah, we do want to set a good example. While I think there's room for nuance about exactly which locator is ideal (the others give you a11y and UX insights), erring on the side of "no flake, please" and using `getByTestId` seems perfectly reasonable.

> This can be revisited in the future, but for now, the punchline is, we have removed flakiness and have a good understanding of what works and what might not.

Nice!

**Comment by Ahmadkashif** [comment]
Hi @ojeytonwilliams @naomi-lgbt :wave: 

I have applied all suggested changes to the PR. 

Let me know if anything else can be done to improve this or else lets get it merged :+1:

**Review by ojeytonwilliams — APPROVED** [review]
Nope, this all looks good to me. Thanks @Ahmadkashif!

**Pull request merged**